### ISSUE TITLE
Invalid PDF structure on shared files

### ISSUE THREAD

**Issue by CamZie** [nextcloud/files_pdfviewer#207]
### Steps to reproduce
1. user shares link of a PDF file
2. opens the PDF shared by link
3. gets error "Invalid or corrupted file | Message: Invalid PDF structure"

### Expected behaviour
shows and can download PDF file through the shared link without any problems

### Actual behaviour
receives error when opening and downloading the PDF file:
![image](https://user-images.githubusercontent.com/16030886/89629631-6b02cf00-d89e-11ea-9994-d9f2378d6e92.png)

The owner of the shared link can access the PDFs so it does not seem to be corrupted. Also re-shared the PDF and still same problem occurs.

### Server configuration

**Operating system:** Debian 9.13 (stretch)

**Web server:** nginx 1.10.3

**Database:** MariaDB 10.1.45

**PHP version:** 7.2

**Nextcloud version:** NC 18.0.6

**Updated from an older Nextcloud/ownCloud or fresh install:** updated from an older version

**Where did you install Nextcloud from:** Nextcloud website

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.4.0
  - activity: 2.11.0
  - admin_audit: 1.8.0
  - apporder: 0.10.0
  - bruteforcesettings: 1.6.0
  - calendar: 2.0.3
  - cloud_federation_api: 1.1.0
  - comments: 1.8.0
  - contacts: 3.3.0
  - dav: 1.14.0
  - drop_account: 0.2.1
  - encryption: 2.6.0
  - external: 3.5.0
  - federatedfilesharing: 1.8.0
  - federation: 1.8.0
  - files: 1.13.1
  - files_pdfviewer: 1.7.0
  - files_rightclick: 0.15.2
  - files_sharing: 1.10.1
  - files_trashbin: 1.8.0
  - files_videoplayer: 1.7.0
  - logreader: 2.3.0
  - lookup_server_connector: 1.6.0
  - notes: 3.5.1
  - notifications: 2.6.0
  - oauth2: 1.6.0
  - password_policy: 1.8.0
  - photos: 1.0.0
  - provisioning_api: 1.8.0
  - quota_warning: 1.7.0
  - recommendations: 0.6.0
  - serverinfo: 1.8.0
  - settings: 1.0.0
  - sharebymail: 1.8.0
  - tasks: 0.13.1
  - text: 2.0.0
  - theming: 1.9.0
  - twofactor_backupcodes: 1.7.0
  - twofactor_totp: 4.1.3
  - updatenotification: 1.8.0
  - viewer: 1.2.0
  - workflowengine: 2.0.0

```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "18.0.6.0",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "forcessl": true,
        "enable_avatars": false,
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "log_authfailip": true,
        "loglevel": 2,
        "maintenance": false,
        "customclient_desktop": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "share_folder": "\/Shared",
        "overwritewebroot": "\/",
        "overwriteprotocol": "https",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "singleuser": false,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "trashbin_retention_obligation": "auto,30",
        "activity_expire_days": 14,
        "logtimezone": "Europe\/Zurich",
        "log_type": "errorlog",
        "skeletondirectory": "\/var\/www\/nextcloud\/themes\/core\/skeleton",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "preview_max_scale_factor": 1,
        "preview_max_filesize_image": 3,
        "preview_max_x": 1024,
        "preview_max_y": 1024,
        "mysql.utf8mb4": true,
        "app_install_overwrite": [
            "drop_account"
        ]
    }
```
</details>

**Are you using encryption:** yes

### Client configuration
**Browser:** Chrome, Firefox

**Operating system:** Linux / Windows

#### Web server error log
<details>
<summary>Web server error log</summary>

```
no errors
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
PHP message: [owncloud][no app in context][3] {"Exception":"OC\\Encryption\\Exceptions\\DecryptionFailedException","Message":"Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.","Code":0,"Trace":[{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":507,"function":"decrypt","class":"OCA\\Encryption\\Crypto\\Encryption","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":302,"function":"readCache","class":"OC\\Files\\Stream\\Encryption","type":"->","args":[]},{"function":"stream_read","class":"OC\\Files\\Stream\\Encryption","type":"->","args":[8192]},{"file":"/var/www/nextcloud/3rdparty/icewind/streams/src/Wrapper.php","line":91,"function":"fread","args":[null,8192]},{"file":"/var/www/nextcloud/3rdparty/icewind/streams/src/CallbackWrapper.php","line":98,"function":"stream_read","class":"Icewind\\Streams\\Wrapper","type":"->","args":[8192" while reading response header from upstream, client: ***REMOVED_SENSITIVE_CONTENT***, server: ***REMOVED_SENSITIVE_CONTENT***, request: "GET /s/tHTmJY64XCQQyr3/download HTTP/2.0", upstream: "fastcgi://unix:/var/run/php-fpm.sock:", host: "***REMOVED_SENSITIVE_CONTENT***"
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
util.js:306 Warning: Ignoring invalid character "33" in hex string
util.js:306 Warning: Ignoring invalid character "79" in hex string
util.js:306 Warning: Ignoring invalid character "84" in hex string
util.js:306 Warning: Ignoring invalid character "89" in hex string
util.js:306 Warning: Ignoring invalid character "80" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "109" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "109" in hex string
2util.js:306 Warning: Ignoring invalid character "108" in hex string
2util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "103" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "112" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "112" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "111" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "114" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "111" in hex string
util.js:306 Warning: Ignoring invalid character "117" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "103" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "111" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "114" in hex string
util.js:306 Warning: Ignoring invalid character "113" in hex string
util.js:306 Warning: Ignoring invalid character "117" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
2util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "111" in hex string
util.js:306 Warning: Ignoring invalid character "107" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
2util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "109" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "114" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "117" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "105" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "33" in hex string
util.js:306 Warning: Ignoring invalid character "79" in hex string
util.js:306 Warning: Ignoring invalid character "84" in hex string
util.js:306 Warning: Ignoring invalid character "89" in hex string
util.js:306 Warning: Ignoring invalid character "80" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "109" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "109" in hex string
2util.js:306 Warning: Ignoring invalid character "108" in hex string
2util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "103" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "112" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "112" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "111" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "114" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "111" in hex string
util.js:306 Warning: Ignoring invalid character "117" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "103" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "111" in hex string
util.js:306 Warning: Ignoring invalid character "108" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Ignoring invalid character "104" in hex string
util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "45" in hex string
util.js:306 Warning: Ignoring invalid character "114" in hex string
util.js:306 Warning: Ignoring invalid character "113" in hex string
util.js:306 Warning: Ignoring invalid character "117" in hex string
util.js:306 Warning: Ignoring invalid character "115" in hex string
2util.js:306 Warning: Ignoring invalid character "116" in hex string
util.js:306 Warning: Ignoring invalid character "111" in hex string
util.js:306 Warning: Ignoring invalid character "107" in hex string
util.js:306 Warning: Ignoring invalid character "110" in hex string
util.js:306 Warning: Ignoring invalid character "61" in hex string
2util.js:306 Warning: Ignoring invalid character "34" in hex string
util.js:306 Warning: Indexing all PDF objects
app.js:692 Uncaught (in promise) Error: Invalid or corrupted PDF file.
    at app.js:692
```
</details>

**Comment by skjnldsv** [comment]
Can you open it properly from regular files? Not from a share?

**Comment by CamZie** [comment]
@skjnldsv Yes, accessing the files normally as the user is working fine. The problem only occurs if the files are opened through a shared link.

**Comment by skjnldsv** [comment]
Have you disabled federated sharing?

**Comment by CamZie** [comment]
We have the following federated sharing enabled:
![image](https://user-images.githubusercontent.com/16030886/91176313-646bb880-e6e2-11ea-9ac1-268b575227fc.png)

**Comment by skjnldsv** [comment]
I will need a screenshot of your network requests.
Before you do anything, still on the development tools, there is a tab called `network`. Click on it and then the pdf GET request. Do your action and screenshot the network log like this:
![capture d ecran_2018-09-27_21-49-46](https://user-images.githubusercontent.com/14975046/46170857-43548000-c29f-11e8-9403-db72713dd9c1.png)

**Comment by skjnldsv** [comment]
If you can also screenshot the preview tab after you clicked the request, it would be nice :)
Thanks!

**Comment by CamZie** [comment]
Thanks! Please let me know if there is anything else you need

![image](https://user-images.githubusercontent.com/16030886/91178667-c2e66600-e6e5-11ea-8959-a523e4592bed.png)

**Comment by skjnldsv** [comment]
Can you click the last one?
Is your pdf 5.5KB?

**Comment by skjnldsv** [comment]
Also, might be fixed in latest nc20, since we moved the pdfviewer :thinking:

**Comment by CamZie** [comment]
This is the preview of the PDF file and it should be 297KB.

![preview](https://user-images.githubusercontent.com/16030886/91185835-96831780-e6ee-11ea-9f83-2b71c2e433de.png)

**Comment by CamZie** [comment]
Also, please take note that this does not only happen for PDF files but also for .zip and photos that are shared through link

> Also, might be fixed in latest nc20, since we moved the pdfviewer thinking

**Comment by hostingnuggets** [comment]
> Also, might be fixed in latest nc20, since we moved the pdfviewer thinking

@skjnldsv could you please confirm that this issue is fixed with Nextcloud 20? Your sentence above only states "might", which is not quite clear... Could you also please backport this fix to Nextcloud 18/19 as I also have this issue and it is affecting a lot of users on my installation too.

**Comment by skjnldsv** [comment]
> > Also, might be fixed in latest nc20, since we moved the pdfviewer thinking
> 
> @skjnldsv could you please confirm that this issue is fixed with Nextcloud 20? Your sentence above only states "might", which is not quite clear... Could you also please backport this fix to Nextcloud 18/19 as I also have this issue and it is affecting a lot of users on my installation too.


I cannot as this is a super edge case and I haven't the time to set this up :)
We upgraded the PDF viewer to latest versions and it now uses the nextcloud Viewer!

**Comment by CamZie** [comment]
@skjnldsv So are there any updates regarding this or any workaround for this issue or do we have to wait until Nextcloud 20 has been released? Will this also solve the problems with the .zip files and images files?


Is it possible that this has to do something with our server side encryption with user keys?

**Comment by skjnldsv** [comment]
I cannot tell. This is out of my league for the encryption part.

What zip and images issue are you referring to?

**Comment by CamZie** [comment]
The issue with the zip and images are the same with the PDF files. The user can open/download it if opened normally but if it is through the shared link it gives the same error as the PDFs. 

> This problem also affects shared files that are .zip. The error appears when trying to download the file:
> 
> > Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.
> 
> The file has also been re-shared and was also re-uploaded but still the same problem.



> This problem also affect images shared through the shared link. It produces the same error as well when trying to download or access the file.
> 
> Is there any news regarding this? This problem is affecting multiple users at the same time...

**Comment by skjnldsv** [comment]
> The issue with the zip and images are the same with the PDF files. The user can open/download it if opened normally but if it is through the shared link it gives the same error as the PDFs. 
> 
> > This problem also affects shared files that are .zip. The error appears when trying to download the file:
> > 
> > > Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.
> > 
> > The file has also been re-shared and was also re-uploaded but still the same problem.
> 
> 
> 
> > This problem also affect images shared through the shared link. It produces the same error as well when trying to download or access the file.
> > 
> > Is there any news regarding this? This problem is affecting multiple users at the same time...

@nextcloud/encryption is it a known issue?

**Comment by hostingnuggets** [comment]
> @nextcloud/encryption is it a known issue?

@skjnldsv note that your team reference to @nextcloud/encryption does not exist... Do you maybe know the people who are directly responsible for the encryption functionality at nextcloud?

**Comment by CamZie** [comment]
Hello everyone, are there any updates regarding the issue with shared files?

**Comment by CamZie** [comment]
We have now upgraded our installation to version 18.0.9 and still the same problem with the shared files... Are there any updates regarding this problem?

**Comment by CamZie** [comment]
Hello, any news regarding this problem?

We have noticed that this mainly affects our installation with server-side encryption with user-keys. Were there any changes regarding the sharing function for this kind of encryption?

**Comment by skjnldsv** [comment]
Sorry, this is unfortunately not in our current priorities :(

**Comment by skjnldsv** [comment]
@CamZie what is the full path of the file?

**Comment by CamZie** [comment]
@skjnldsv Below you will find the browser log (with the path) when trying to access the PDF file in a shared link:

![network_browser](https://user-images.githubusercontent.com/16030886/98531402-940d7400-2280-11eb-9fac-508edc8bb3e5.png)

**Comment by skjnldsv** [comment]
Can you share a link with me so I can investigate on your instance please?
john@nextcloud.com

**Comment by CamZie** [comment]
@skjnldsv I have now sent an email with the shared links to john@nextcloud.com . Thanks in advance for the help.

**Comment by skjnldsv** [comment]
Yeah, turns out this is not possible :thinking: 
https://docs.nextcloud.com/server/stable/user_manual/en/files/encrypting_files.html#sharing-encrypted-files

**Comment by yahesh** [comment]
> Yeah, turns out this is not possible thinking
> https://docs.nextcloud.com/server/stable/user_manual/en/files/encrypting_files.html#sharing-encrypted-files

I'd argue against that. Nextcloud allows to publicly share SSE encrypted files by means of the pubShare keypair. The link you provided relates to user key encryption iin cases where not all user keypairs have been generated. However, the pubShare keypair should be generated as soon as the SSE is enabled.

**Comment by skjnldsv** [comment]
I have no clue to be honest, but it seems that it is disabled.
This is not an issue about pdfviewer, downloading is impossible as well.

**Comment by yahesh** [comment]
@CamZie Could have a look into your data directory whether you have a file called `files_encryption/OC_DEFAULT_MODULE/pubShare_$random.publicKey` in there where `$random` is a random identifier? After you shared a file through a link, could you also have a look whether you can find a file called `$username/files_encryption/keys/files/$filename/OC_DEFAULT_MODULE/pubShare_$random.shareKey` where `$username` is the owner of the shared file, `$filename` is name of the shared file and `$random` is the same random identifier as before?

**Comment by CamZie** [comment]
@yahesh As requested, I confirm that the files are existing and the random identifier of the keys are the same, which i replaced here with `abcd1234`:

```
data/files_encryption/OC_DEFAULT_MODULE/pubShare_abcd1234.privateKey  
data/files_encryption/OC_DEFAULT_MODULE/pubShare_abcd1234.publicKey

<USERNAME>/files_encryption/keys/files/document.pdf/OC_DEFAULT_MODULE/pubShare_abcd1234.shareKey
```

**Comment by yahesh** [comment]
Then Nextcloud should be able to decrypt the file when it is accessed through the share link.

**Comment by skjnldsv** [comment]
I'm out of ideas, does it work for you @yahesh ?

**Comment by ozboss** [comment]
So I noticed I get this error if I try to open a pdf file in a folder with '#' prepended (e.g. /#testfolder). I can view the pdf without issues if it is placed in a folder like /testfolder. So special characters in the link / folder structure are causing this???

**Comment by yahesh** [comment]
> So I noticed I get this error if I try to open a pdf file in a folder with '#' prepended (e.g. /#testfolder). I can view the pdf without issues if it is placed in a folder like /testfolder. So special characters in the link / folder structure are causing this???

Gernally folder and file names names containing hash signs (`#`) and question marks (`?`) seem to break the PDF viewer. I assume that this stems from some kind of escaping problem as both characters have special meanings in URLs.

**Comment by skjnldsv** [comment]
https://github.com/nextcloud/viewer/issues/415

**Comment by CamZie** [comment]
We are also aware of the problems with folder/file names containing special characters. But this is not cause of our issue because we are not using any special characters in the names.

**Comment by JBM888** [comment]
Hello,
I'm experiancing the same problem having PDF's getting corrupted when trying to view them as shared with read only no download rights. The strange thing is that it worked initially but then stoped. I have tried to get it to work again by moving the firectory out of its initial path since that had a character that, according to reasoning earlier in the thread, could be an issue. 
I'm currently running NextCloud 20.0.7 which appears healthy on a Raspberry.

Regards! 
/JBM

**Comment by baltaner** [comment]
Hello,
I've nearly the same problems here.
Thoe problem occur ONLY when the PDF-File is in an public shared folder with password protection (enforced). Whene I download this file the file opens without any problems!

When I share the same PDF-File directly (not inside a shared folder) it works as expected - I get the correct view of the PDF-File inside the browser!

I hope this helps to isolate the problem.

Best regards
Bertold

**Comment by nderambure** [comment]
Hello,

Same problem here, but with a variation : I have a folder with a PDF inside.
If I'm connected to my account and click the PDF : no problem, the PDF is previewed.
But if I share the folder with a public link, and then go to this link and try to open the PDF, it fails with the error described above.

![screenshot_error_pdfviewer](https://user-images.githubusercontent.com/1145642/107249344-93c57600-6a33-11eb-86fe-c66b08789089.png)

**Comment by bigahuna** [comment]
Steps to reproduce
- user shares link of a folder containing multiple PDF files
- opens the folder shared by link
- opens a pdf file
- gets error "Invalid or corrupted file | Message: Invalid PDF structure"
- download of the file works, the file can then be opened in multiple pdf viewers locally (Acrobat reader, Macos preview, google chrome, firefox)

Expected behaviour
- shows and can download PDF file through the shared link without any problems

Actual behaviour
After Updating from nextcloud 20.0.6 to 20.0.7 a click on a pdf file in a shared folder produces the following error:

<img width="332" alt="Bildschirmfoto 2021-02-10 um 13 29 02" src="https://user-images.githubusercontent.com/13297945/107510289-fe95bf00-6ba3-11eb-939a-4b8dd7448bf1.png">

The same files can be opened without problems inside non shared folders. This is reproducable.

Server configuration
Operating system: Centos 7 latest
Web server: Apache/2.4.6 (CentOS)
Database: MariaDB 10.3.27
PHP version: 7.4.14
Nextcloud version: NC  20.0.7

**Comment by dst21** [comment]
Same here after Update to Nextcloud 20.0.7.> Message: Invalid PDF structure. 

In the meantime I will set Onlyoffice as the default PDF viewer.

**Comment by bigahuna** [comment]
That doesn't work either. Opening PDFs with OnlyOffice in **the shared folder** results in this error:
<img width="682" alt="Bildschirmfoto 2021-02-10 um 14 32 14" src="https://user-images.githubusercontent.com/13297945/107516555-df4f5f80-6bac-11eb-9eaa-54c78d2222ab.png">

Open PDFs with OnlyOffice in a non shared folder works just fine. The Path to the PDF files is corrupted inside shared folders, so the app to view them does not matter because they can't be opened-

**Comment by dst21** [comment]
This is peculiar. For me, the view works in a shared folder with OnlyOffice

**Comment by misterscrooby** [comment]
Same problem here in Nextcloud 20.0.7. It only occoures in shared folder via links.

**Comment by nderambure** [comment]
I seems @wankdanker has a patch for this, Kudos to him : https://github.com/nextcloud/files_pdfviewer/pull/309

**Comment by nderambure** [comment]
Patch has been made, adapting original one, and seems to be delivered soon in stable release, see https://github.com/nextcloud/files_pdfviewer/pull/319

**Comment by ThinkOli** [comment]
I HAD the same problem. Resolved with the help of this discussion ...

We ran server side encryption and ran into some issues with accessing shared files. Given we're running encryption on the filesystem, we decided to abandon NC encryption and I ran thru the process:
`cd /var/www/html/nextcloud/`
`sudo -u www-data php occ encryption:decrypt-all`
`sudo -u www-data php occ maintenance:mode --on`
`sudo -u www-data php occ encryption:disable`
`sudo -u www-data php occ maintenance:mode --off`

Unfortunately, not all files got decrypted. I had to learn that a lot of data remained encrypted. But status of encryption was 'enabled: false'. I.e. I could not simply rerun the decryption. After
`sudo -u www-data php occ maintenance:mode --on`
`sudo -u www-data php occ encryption:enable`
`sudo -u www-data php occ maintenance:mode --off`
`sudo -u www-data php occ encryption:decrypt-all`
upon successfull completion of the last step (after several hours), encryption got disabled again, and all data became accessible.

I have to say that I ssh'ed into the server and experienced some disconnections, i.e. the process could not be completed within a single step. With every reconnect, encryption had got disabled, and I had to start the above procedure. But when it came back with the system prompt, everything is now working well.

Whether we'll decide on leveraging server side encryption again in the future, remains to be seen. So far, for us it did not work reliably.

**Comment by rmdes** [comment]
I have tried some of the advices listed in this thread but none of them could help me get at the same time, sharing to work, PDF to embed and files to be downloadable. 

The issue is clearly the PDFjs viewer because the moment this app is disabled, files can be downloaded just fine and sharing link works perfectly. 

**So this is my temporary fix for this bug, disable PDF Viewer**

Version : 
 Nextcloud 24.0.4 
 files not encrypted
 basically a default Nextcloud.
